 <h1>Fifa Data Cleaning Challenge </h1>

<p> Data Cleaning is a very important part of data analysis. Clean Data ensures that <br>
your analysis is correct and your insights are reliable when used in decision <br>
making. </p>

<h2> Objectives: </h2>
<p> - Ensure that all columns are clearly named</p>
<p> - Ensure that columns have the correct datatypes</p>
<p> - Remove all unnecessary information from the dataset</p>

<p> <b> Lets Get Started: </p>

<h2> Import all the necessary libraries needed for this project </h2> 

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

<h3> import the csv data set from the folder where its saved </h3>

In [3]:
df = pd.read_csv(r'C:\Users\Admin\Desktop\csv files\fifa21 raw data v2.csv')

<h3> set your dataframe such that it displays all columns since the dataset <br>
has many columns. The benefit is that not only all colums are well displayed,<br>
but also the printed rows can be larger than the usual ~100 characters limit. </h3>

In [5]:
pd.set_option('display.max_rows', 1000)
pd.set_option('display.max_columns', 1000)
pd.set_option('display.width', 1000)

<h3> Display the first five rows </h3>

In [7]:
df.head()

ID               Name                      LongName                                          photoUrl                                          playerUrl Nationality  Age  ↓OVA  POT                         Club     Contract   Positions Height Weight Preferred Foot  BOV Best Position     Joined Loan Date End    Value   Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W     D/W   IR  PAC  SHO  PAS  DRI  DEF  PHY Hits
0  158023           L. Messi                  Lionel Messi  https://cdn.sofifa.com/players/158/023/21_60.png  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg           Left   93            RW  01-Jul-04           NaN  €103.5M  €560K        €138.4M        429        85         95                70             91       88    470         96     93           94            91            96       451            91            80       91         94       95    389          86       68       72        69          94        347          44             40           93      95         75         96         91       32               35              24           54          6           11          15              14            8         2231         466  4 ★  4★  Medium     Low  5 ★   85   92   91   95   38   65  771
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro  https://cdn.sofifa.com/players/020/801/21_60.png  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg          Right   92            ST  10-Jul-18           NaN     €63M  €220K         €75.9M        437        84         95                90             82       86    414         88     81           76            77            92       431            87            91       87         95       71    444          94       95       84        78          93        353          63             29           95      82         84         95         84       28               32              24           58          7           11          15              14           11         2221         464  4 ★  5★    High     Low  5 ★   89   93   81   89   35   77  562
2  200389           J. Oblak                     Jan Oblak  https://cdn.sofifa.com/players/200/389/21_60.png  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg          Right   91            GK  16-Jul-14           NaN    €120M  €125K        €159.4M         95        13         11                15             43       13    109         12     13           14            40            30       307            43            60       67         88       49    268          59       78       41        78          12        140          34             19           11      65         11         68         57       27               12              18          437         87           92          78              90           90         1413         489  3 ★  1★  Medium  Medium  3 ★   87   92   78   90   52   90  150
3  192985       K. De Bruyne               Kevin De Bruyne  https://cdn.sofifa.com/players/192/985/21_60.png  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg          Right   91           CAM  30-Aug-15           NaN    €129M  €370K          €161M      

<h3> Check how many rows and columns make up your dafaframe</h3>

In [9]:
df.shape

(18979, 77)

In [11]:
print('The number of rows are:', df.shape[0])
print('The number of columns are:', df.shape[1])

The number of rows are: 18979
The number of columns are: 77


<h3> Get summary information from your dataset.</h3>

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

<p> 1. From the above information the dataset has 18979 rows and 77 columns <br>
2. Most columns dont have null values except 2 columns that is:<br>
- Loan Date End<br>
- Hits. <br>
<b>3. Also the columns are made up of two data types: </b><br>
- object<br>
- int64</p>

<h2> Column Names: </h2>
<P> Display a list of all column names and iterate through<br>
    to see columns of your data frame</P>

In [17]:
for x in df.columns.tolist():
    print(x)

ID
Name
LongName
photoUrl
playerUrl
Nationality
Age
↓OVA
POT
Club
Contract
Positions
Height
Weight
Preferred Foot
BOV
Best Position
Joined
Loan Date End
Value
Wage
Release Clause
Attacking
Crossing
Finishing
Heading Accuracy
Short Passing
Volleys
Skill
Dribbling
Curve
FK Accuracy
Long Passing
Ball Control
Movement
Acceleration
Sprint Speed
Agility
Reactions
Balance
Power
Shot Power
Jumping
Stamina
Strength
Long Shots
Mentality
Aggression
Interceptions
Positioning
Vision
Penalties
Composure
Defending
Marking
Standing Tackle
Sliding Tackle
Goalkeeping
GK Diving
GK Handling
GK Kicking
GK Positioning
GK Reflexes
Total Stats
Base Stats
W/F
SM
A/W
D/W
IR
PAC
SHO
PAS
DRI
DEF
PHY
Hits


<h3> Make a copy of your data set so as to retain an original copy</h3>

In [20]:
df1 = df.copy()

In [22]:
df1.head()

ID               Name                      LongName                                          photoUrl                                          playerUrl Nationality  Age  ↓OVA  POT                         Club     Contract   Positions Height Weight Preferred Foot  BOV Best Position     Joined Loan Date End    Value   Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W     D/W   IR  PAC  SHO  PAS  DRI  DEF  PHY Hits
0  158023           L. Messi                  Lionel Messi  https://cdn.sofifa.com/players/158/023/21_60.png  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg           Left   93            RW  01-Jul-04           NaN  €103.5M  €560K        €138.4M        429        85         95                70             91       88    470         96     93           94            91            96       451            91            80       91         94       95    389          86       68       72        69          94        347          44             40           93      95         75         96         91       32               35              24           54          6           11          15              14            8         2231         466  4 ★  4★  Medium     Low  5 ★   85   92   91   95   38   65  771
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro  https://cdn.sofifa.com/players/020/801/21_60.png  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg          Right   92            ST  10-Jul-18           NaN     €63M  €220K         €75.9M        437        84         95                90             82       86    414         88     81           76            77            92       431            87            91       87         95       71    444          94       95       84        78          93        353          63             29           95      82         84         95         84       28               32              24           58          7           11          15              14           11         2221         464  4 ★  5★    High     Low  5 ★   89   93   81   89   35   77  562
2  200389           J. Oblak                     Jan Oblak  https://cdn.sofifa.com/players/200/389/21_60.png  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg          Right   91            GK  16-Jul-14           NaN    €120M  €125K        €159.4M         95        13         11                15             43       13    109         12     13           14            40            30       307            43            60       67         88       49    268          59       78       41        78          12        140          34             19           11      65         11         68         57       27               12              18          437         87           92          78              90           90         1413         489  3 ★  1★  Medium  Medium  3 ★   87   92   78   90   52   90  150
3  192985       K. De Bruyne               Kevin De Bruyne  https://cdn.sofifa.com/players/192/985/21_60.png  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg          Right   91           CAM  30-Aug-15           NaN    €129M  €370K          €161M      

<h3> Remove unnecessary columns</h3>

In [24]:
df1 = df1.drop(['Name', 'photoUrl', 'playerUrl'], axis = 1)

In [26]:
df1.head(1)

ID      LongName Nationality  Age  ↓OVA  POT                  Club     Contract   Positions Height Weight Preferred Foot  BOV Best Position     Joined Loan Date End    Value   Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  PAS  DRI  DEF  PHY Hits
0  158023  Lionel Messi   Argentina   33    93   93  \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg           Left   93            RW  01-Jul-04           NaN  €103.5M  €560K        €138.4M        429        85         95                70             91       88    470         96     93           94            91            96       451            91            80       91         94       95    389          86       68       72        69          94        347          44             40           93      95         75         96         91       32               35              24           54          6           11          15              14            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   91   95   38   65  771

<h3>Rename columns</h3>

In [28]:
df1 = df1.rename(columns = {
                "LongName": "Name",
                "↓OVA":"Overall Rating(%)", 
                 "POT": "Potential(%)",
                 "BOV": "Best Overall(%)",
                 "BP": "Best Position",
                 "W/F": "Weak Foot",
                 "SM": "Skill Moves",
                 "A/W": "Attacking Work Rate",
                 "D/W": "Defensive Work Rate",
                 "IR": "International Reputation",
                 "PAC": "Pace",
                 "SHO": "Shooting",
                 "PAS": "Passing",
                 "DRI": "Dribbling",
                 "DEF": "Defense",
                 "PHY": "Physicality" 
})

<h3> Check for duplicates</h3>

In [33]:
df1['Name'].nunique()

18852

<p> Out of 18979 rows there are 18852 unique rows</p>

In [35]:
duplicated_rows = df1[df1.duplicated(["Name"])]
duplicated_rows.head(2)

ID            Name Nationality  Age  Overall Rating(%)  Potential(%)                       Club     Contract Positions Height Weight Preferred Foot  Best Overall(%) Best Position     Joined Loan Date End  Value  Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats Weak Foot Skill Moves Attacking Work Rate Defensive Work Rate International Reputation  Pace  Shooting  Passing  Dribbling  Defense  Physicality Hits
1239  215051  Lisandro López   Argentina   30                 76            76       \n\n\n\nBoca Juniors  2020 ~ 2023        CB  188cm   80kg          Right               76            CB  01-Jan-20           NaN    €6M  €16K          €8.4M        277        33         48                80             66       50    246         59     37           32            56            62       310            57            66       64         73       50    335          67       84       63        77          44        279          77             77           38      41         46         73        229       76               77              76           53         14           12           8               6           13         1729         375       2 ★          2★              Medium                High                      3 ★    62        51       50         61       77           74   17
2511  213017      Ben Davies     England   24                 73            79  \n\n\n\nPreston North End  2013 ~ 2021        CB  185cm   74kg           Left               75            CB  01-Jan-13           NaN  €4.4M  €18K         €10.3M        283        61         36                72             71       43    277         59     44           40            68            66       348            75            73       65         70       65    312          42       78       75        78          39        256          65             72           37      34         48         71        225       79               74              72           52          6           15          10              15            6         1753         384       3 ★          2★              Medium              Medium                      1 ★    74        39       58         63       75           75   30

<p> I investigated one of the values to see if truly they are duplicates but<br/>
based on the findings below they are not duplicated values they only have<br/>
similar names.</p>

In [37]:
filt = df[df1.Name == "Ben Davies"]
filt

ID       Name    LongName                                          photoUrl                                          playerUrl Nationality  Age  ↓OVA  POT                       Club     Contract Positions Height Weight Preferred Foot  BOV Best Position     Joined Loan Date End  Value  Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W     D/W   IR  PAC  SHO  PAS  DRI  DEF  PHY Hits
382   205923  B. Davies  Ben Davies  https://cdn.sofifa.com/players/205/923/21_60.png  http://sofifa.com/player/205923/ben-davies/210...       Wales   27    80   81  \n\n\n\nTottenham Hotspur  2014 ~ 2024   LB, LWB  182cm   76kg           Left   80            LB  23-Jul-14           NaN   €22M  €86K         €28.6M        329        78         58                66             79       48    343         75     72           48            70            78       360            73            73       68         78       68    343          65       74       83        69          52        348          76             79           64      75         54         74        239       77               82              80           45         11            7           6               7           14         2007         433  3 ★  3★  Medium    High  2 ★   73   58   75   75   78   74   64
2511  213017  B. Davies  Ben Davies  https://cdn.sofifa.com/players/213/017/21_60.png  http://sofifa.com/player/213017/ben-davies/210...     England   24    73   79  \n\n\n\nPreston North End  2013 ~ 2021        CB  185cm   74kg           Left   75            CB  01-Jan-13           NaN  €4.4M  €18K         €10.3M        283        61         36                72             71       43    277         59     44           40            68            66       348            75            73       65         70       65    312          42       78       75        78          39        256          65             72           37      34         48         71        225       79               74              72           52          6           15          10              15            6         1753         384  3 ★  2★  Medium  Medium  1 ★   74   39   58   63   75   75   30

<p>Remove whitespaces in the club column </p>

In [40]:
df1['Club'].head()

0           \n\n\n\nFC Barcelona
1               \n\n\n\nJuventus
2        \n\n\n\nAtlético Madrid
3        \n\n\n\nManchester City
4    \n\n\n\nParis Saint-Germain
Name: Club, dtype: object

In [42]:
#Remove whitespaces from the club column
df1['Club'] = df1['Club'].str.lstrip()

In [44]:
df1.head(2)

ID                          Name Nationality  Age  Overall Rating(%)  Potential(%)          Club     Contract   Positions Height Weight Preferred Foot  Best Overall(%) Best Position     Joined Loan Date End    Value   Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats Weak Foot Skill Moves Attacking Work Rate Defensive Work Rate International Reputation  Pace  Shooting  Passing  Dribbling  Defense  Physicality Hits
0  158023                  Lionel Messi   Argentina   33                 93            93  FC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg           Left               93            RW  01-Jul-04           NaN  €103.5M  €560K        €138.4M        429        85         95                70             91       88    470         96     93           94            91            96       451            91            80       91         94       95    389          86       68       72        69          94        347          44             40           93      95         75         96         91       32               35              24           54          6           11          15              14            8         2231         466       4 ★          4★              Medium                 Low                      5 ★    85        92       91         95       38           65  771
1   20801  C. Ronaldo dos Santos Aveiro    Portugal   35                 92            92      Juventus  2018 ~ 2022      ST, LW  187cm   83kg          Right               92            ST  10-Jul-18           NaN     €63M  €220K         €75.9M        437        84         95                90             82       86    414         88     81           76            77            92       431            87            91       87         95       71    444          94       95       84        78          93        353          63             29           95      82         84         95         84       28               32              24           58          7           11          15              14           11         2221         464       4 ★          5★                High                 Low                      5 ★    89        93       81         89       35           77  562

<p>Check for unique values in the contract column</p>

In [47]:
df1['Contract'].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

In [49]:
#define a function to change contract column values
def contract_status(value):
    if 'On Loan' in value:
        value = 'On Loan'
        return value
    elif '~' in value:
        value = 'Active'
        return value
    else:
        value = 'Free'
        return value

In [51]:
#apply the function on contract column
df1['Contract'] = df1['Contract'].apply(contract_status).astype('category')

In [53]:
#Check for ujnique values in the contract column
df1['Contract'].unique()

['Active', 'On Loan', 'Free']
Categories (3, object): ['Active', 'Free', 'On Loan']

In [55]:
#Rename the contract column
df1 = df1.rename(columns = {'Contract': 'Contract Status'})
df1.head(1)

ID          Name Nationality  Age  Overall Rating(%)  Potential(%)          Club Contract Status   Positions Height Weight Preferred Foot  Best Overall(%) Best Position     Joined Loan Date End    Value   Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats Weak Foot Skill Moves Attacking Work Rate Defensive Work Rate International Reputation  Pace  Shooting  Passing  Dribbling  Defense  Physicality Hits
0  158023  Lionel Messi   Argentina   33                 93            93  FC Barcelona          Active  RW, ST, CF  170cm   72kg           Left               93            RW  01-Jul-04           NaN  €103.5M  €560K        €138.4M        429        85         95                70             91       88    470         96     93           94            91            96       451            91            80       91         94       95    389          86       68       72        69          94        347          44             40           93      95         75         96         91       32               35              24           54          6           11          15              14            8         2231         466       4 ★          4★              Medium                 Low                      5 ★    85        92       91         95       38           65  771

<h3>Process of cleaning contract column</h3>
<P>
A function was defined to change the row values from '2004 ~ 2021', 'On Loan' and 'Free' to 'Active', 'On Loan' and 'Free'<br>
'Contract' column was renamed to 'Contract Status<br>.
Also, the data type was changed to categoy.</Pory.

<p>Inspect Position and best position columns</p>

In [57]:
#Position and best position
df1[['Positions', 'Best Position']].head()

Positions Best Position
0  RW, ST, CF            RW
1      ST, LW            ST
2          GK            GK
3     CAM, CM           CAM
4     LW, CAM            LW

<p>From the above findings Positions holds the players best position and <br>
other positions held by the player so i dropped positions column.</p>

In [60]:
#drop the position column
df1 = df1.drop('Positions', axis = 1)
df1.head(2)

ID                          Name Nationality  Age  Overall Rating(%)  Potential(%)          Club Contract Status Height Weight Preferred Foot  Best Overall(%) Best Position     Joined Loan Date End    Value   Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats Weak Foot Skill Moves Attacking Work Rate Defensive Work Rate International Reputation  Pace  Shooting  Passing  Dribbling  Defense  Physicality Hits
0  158023                  Lionel Messi   Argentina   33                 93            93  FC Barcelona          Active  170cm   72kg           Left               93            RW  01-Jul-04           NaN  €103.5M  €560K        €138.4M        429        85         95                70             91       88    470         96     93           94            91            96       451            91            80       91         94       95    389          86       68       72        69          94        347          44             40           93      95         75         96         91       32               35              24           54          6           11          15              14            8         2231         466       4 ★          4★              Medium                 Low                      5 ★    85        92       91         95       38           65  771
1   20801  C. Ronaldo dos Santos Aveiro    Portugal   35                 92            92      Juventus          Active  187cm   83kg          Right               92            ST  10-Jul-18           NaN     €63M  €220K         €75.9M        437        84         95                90             82       86    414         88     81           76            77            92       431            87            91       87         95       71    444          94       95       84        78          93        353          63             29           95      82         84         95         84       28               32              24           58          7           11          15              14           11         2221         464       4 ★          5★                High                 Low                      5 ★    89        93       81         89       35           77  562

<h3>Check Height and Weight Columns</h3>

In [63]:
#Check for unique values
for column in df[['Height', 'Weight']]:
    value = df1[column].unique()
    print(f'{column}\n{value}.\n')

Height
['170cm' '187cm' '188cm' '181cm' '175cm' '184cm' '191cm' '178cm' '193cm'
 '185cm' '199cm' '173cm' '168cm' '176cm' '177cm' '183cm' '180cm' '189cm'
 '179cm' '195cm' '172cm' '182cm' '186cm' '192cm' '165cm' '194cm' '167cm'
 '196cm' '163cm' '190cm' '174cm' '169cm' '171cm' '197cm' '200cm' '166cm'
 '6\'2"' '164cm' '198cm' '6\'3"' '6\'5"' '5\'11"' '6\'4"' '6\'1"' '6\'0"'
 '5\'10"' '5\'9"' '5\'6"' '5\'7"' '5\'4"' '201cm' '158cm' '162cm' '161cm'
 '160cm' '203cm' '157cm' '156cm' '202cm' '159cm' '206cm' '155cm'].

Weight
['72kg' '83kg' '87kg' '70kg' '68kg' '80kg' '71kg' '91kg' '73kg' '85kg'
 '92kg' '69kg' '84kg' '96kg' '81kg' '82kg' '75kg' '86kg' '89kg' '74kg'
 '76kg' '64kg' '78kg' '90kg' '66kg' '60kg' '94kg' '79kg' '67kg' '65kg'
 '59kg' '61kg' '93kg' '88kg' '97kg' '77kg' '62kg' '63kg' '95kg' '100kg'
 '58kg' '183lbs' '179lbs' '172lbs' '196lbs' '176lbs' '185lbs' '170lbs'
 '203lbs' '168lbs' '161lbs' '146lbs' '130lbs' '190lbs' '174lbs' '148lbs'
 '165lbs' '159lbs' '192lbs' '181lbs' '139lbs' '15

In [65]:
#Function to convert height to cm 
def convert_height(value):
    if 'cm' in value:
        value = int(value[:-2])
        return value
    else:
        feet, inches = value.split("'")
        total_inches = int(feet) * 12 + int(inches[:-1]) 
        height_cm = total_inches * 2.54
        return round(height_cm, 2)

df1['Height'] = df1['Height'].apply(convert_height).astype('int64')
df1['Height'].unique()

array([170, 187, 188, 181, 175, 184, 191, 178, 193, 185, 199, 173, 168,
       176, 177, 183, 180, 189, 179, 195, 172, 182, 186, 192, 165, 194,
       167, 196, 163, 190, 174, 169, 171, 197, 200, 166, 164, 198, 162,
       201, 158, 161, 160, 203, 157, 156, 202, 159, 206, 155], dtype=int64)

In [67]:
#function to convert weight to kg
def convert_weight(value):
    if 'kg' in value:
        value = value.strip('kg')
        return value
    else:
        value = value.strip('lbs')
        Weight = round((float(value) * 0.45359237), 2)
        return Weight    

df1['Weight'] = df1['Weight'].apply(convert_weight).astype('int64')
df1['Weight'].unique()

array([ 72,  83,  87,  70,  68,  80,  71,  91,  73,  85,  92,  69,  84,
        96,  81,  82,  75,  86,  89,  74,  76,  64,  78,  90,  66,  60,
        94,  79,  67,  65,  59,  61,  93,  88,  97,  77,  62,  63,  95,
       100,  58,  98, 103,  99, 102,  56, 101,  57,  55, 104, 107, 110,
        53,  50,  54,  52], dtype=int64)

In [69]:
#Rename Height and Weight columns
df1 = df1.rename(columns = {'Height' : 'Height(cm)', 'Weight' : 'Weight(kg)'})
df1.head()

ID                          Name Nationality  Age  Overall Rating(%)  Potential(%)                 Club Contract Status  Height(cm)  Weight(kg) Preferred Foot  Best Overall(%) Best Position     Joined Loan Date End    Value   Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats Weak Foot Skill Moves Attacking Work Rate Defensive Work Rate International Reputation  Pace  Shooting  Passing  Dribbling  Defense  Physicality Hits
0  158023                  Lionel Messi   Argentina   33                 93            93         FC Barcelona          Active         170          72           Left               93            RW  01-Jul-04           NaN  €103.5M  €560K        €138.4M        429        85         95                70             91       88    470         96     93           94            91            96       451            91            80       91         94       95    389          86       68       72        69          94        347          44             40           93      95         75         96         91       32               35              24           54          6           11          15              14            8         2231         466       4 ★          4★              Medium                 Low                      5 ★    85        92       91         95       38           65  771
1   20801  C. Ronaldo dos Santos Aveiro    Portugal   35                 92            92             Juventus          Active         187          83          Right               92            ST  10-Jul-18           NaN     €63M  €220K         €75.9M        437        84         95                90             82       86    414         88     81           76            77            92       431            87            91       87         95       71    444          94       95       84        78          93        353          63             29           95      82         84         95         84       28               32              24           58          7           11          15              14           11         2221         464       4 ★          5★                High                 Low                      5 ★    89        93       81         89       35           77  562
2  200389                     Jan Oblak    Slovenia   27                 91            93      Atlético Madrid          Active         188          87          Right               91            GK  16-Jul-14           NaN    €120M  €125K        €159.4M         95        13         11                15             43       13    109         12     13           14            40            30       307            43            60       67         88       49    268          59       78       41        78          12        140          34             19           11      65         11         68         57       27               12              18          437         87           92          78              90           90         1413         489       3 ★          1★              Medium              Medium                      3 ★    87        92       78         90       52           90  150
3  192985               Kevin De Bruyne     Belgium   29                 91            91      Manchester City          Active         181          70          Right               91           CAM  30-Aug-15           NaN    €129M  €370K          €161M        407        94         82                55             94       82    441         88     85           83            93     

<h3>Process of Cleaning 'Height' and 'Weight' columns</h3>
<p>1. Create a function to convert Height <br>
2. values with cm remain the same except that we extract only value without 'CM'<br>
3. value in feet and inches we convert them to CM <br>
4. Lastly, convert Height column to int</p>
<h3>Process of cleaning Weight column <br></h3>
<p>1. create a function to convert LBS to kgs<br>
2. If value is in "KGS" it remains the same <br>
3. If value is in "LBS" convert them to KGS by multiplying by 0.4535...</p>

<h3>Convert date columns to date time using pandas</h3>

In [71]:
for column in df[['Joined', 'Loan Date End']]:
    value = df1[column].unique()
    print(f'{column}\n{value}\n')

Joined
['01-Jul-04' '10-Jul-18' '16-Jul-14' ... '22-Sep-18' '28-Feb-15'
 '06-Mar-18']

Loan Date End
[nan '30-Jun-21' '31-Dec-20' '30-Jan-21' '30-Jun-22' '31-May-21'
 '05-Jul-21' '31-Dec-21' '01-Jul-21' '01-Jan-21' '31-Aug-21' '31-Jan-21'
 '30-Dec-21' '23-Jun-21' '03-Jan-21' '27-Nov-21' '17-Jan-21' '30-Jun-23'
 '31-Jul-21' '22-Nov-20' '31-May-22' '30-Dec-20' '04-Jan-21' '30-Nov-20'
 '01-Aug-21']



In [73]:
#Covert date columns to datetime data types
df1['Joined'] = pd.to_datetime(df1['Joined'])

In [75]:
df1['Loan Date End'] = pd.to_datetime(df1['Loan Date End'])

In [77]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 73 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   ID                        18979 non-null  int64         
 1   Name                      18979 non-null  object        
 2   Nationality               18979 non-null  object        
 3   Age                       18979 non-null  int64         
 4   Overall Rating(%)         18979 non-null  int64         
 5   Potential(%)              18979 non-null  int64         
 6   Club                      18979 non-null  object        
 7   Contract Status           18979 non-null  category      
 8   Height(cm)                18979 non-null  int64         
 9   Weight(kg)                18979 non-null  int64         
 10  Preferred Foot            18979 non-null  object        
 11  Best Overall(%)           18979 non-null  int64         
 12  Best Position     

<h3>Look at "Value", "Wage", "Release Clause" columns</h3>

In [80]:
df1[["Value", "Wage", "Release Clause"]].head(4)


Value   Wage Release Clause
0  €103.5M  €560K        €138.4M
1     €63M  €220K         €75.9M
2    €120M  €125K        €159.4M
3    €129M  €370K          €161M

In [82]:
# Creating function to convert string to numeric
def convert_to_numeric(value):
    #Remove currency symbol
    value_str = value.replace('€', '').replace(',', '')
    if 'K' in value_str:
        return float(value_str.replace('K', '')) * 1000
    elif 'M' in value_str:
        return float(value_str.replace('M', '')) * 1000000
    else:
        return float(value_str)
# Apply the conversion function to the "Wage" and "Value" columns
df1["Wage"] = df1["Wage"].apply(convert_to_numeric)
df1["Value"] = df1["Value"].apply(convert_to_numeric)
df1["Release Clause"] = df1["Release Clause"].apply(convert_to_numeric)



In [84]:
#display the first five rows
df1[["Value", "Wage", "Release Clause"]].head()

Value      Wage  Release Clause
0  103500000.0  560000.0     138400000.0
1   63000000.0  220000.0      75900000.0
2  120000000.0  125000.0     159400000.0
3  129000000.0  370000.0     161000000.0
4  132000000.0  270000.0     166500000.0

<h3>Process</h3>
<p>1. Create a function to convert values to numeric <br>
2. Remove '€' and ',' from the columns <br>
3. Check for 'K' in values and multiply by 1000 <br>
4. Check for 'M' in value and multiply by 1000000 <br>
5. Apply the function to those columns</p>

### Inspect 'Weak Foot', 'Skill Moves', 'International Reputation' Columns

In [86]:
df1.head(1)

ID          Name Nationality  Age  Overall Rating(%)  Potential(%)          Club Contract Status  Height(cm)  Weight(kg) Preferred Foot  Best Overall(%) Best Position     Joined Loan Date End        Value      Wage  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  Composure  Defending  Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats Weak Foot Skill Moves Attacking Work Rate Defensive Work Rate International Reputation  Pace  Shooting  Passing  Dribbling  Defense  Physicality Hits
0  158023  Lionel Messi   Argentina   33                 93            93  FC Barcelona          Active         170          72           Left               93            RW 2004-07-01           NaT  103500000.0  560000.0     138400000.0        429        85         95                70             91       88    470         96     93           94            91            96       451            91            80       91         94       95    389          86       68       72        69          94        347          44             40           93      95         75         96         91       32               35              24           54          6           11          15              14            8         2231         466       4 ★          4★              Medium                 Low                      5 ★    85        92       91         95       38           65  771

In [88]:
for column in df1[['Weak Foot', 'Skill Moves', 'International Reputation']]:
    value = df1[column].unique()
    print(f'{column}\n{value}\n')

Weak Foot
['4 ★' '3 ★' '5 ★' '2 ★' '1 ★']

Skill Moves
['4★' '5★' '1★' '2★' '3★']

International Reputation
['5 ★' '3 ★' '4 ★' '2 ★' '1 ★']



In [90]:
def convert_to_num(value):
    value = int(value[:-1])
    return value

df1['Weak Foot'] = df1['Weak Foot'].apply(convert_to_num).astype('int')

In [92]:
df1['Skill Moves'] = df1['Skill Moves'].apply(convert_to_num).astype('int')
df1['International Reputation'] = df1['International Reputation'].apply(convert_to_num).astype('int')

<h3>Process</h3>
<p>Created a Function that only extracts value leaving out <br>
the last character which is a special character '★'<br>
After removing the special character i converted the column to int data type</p>

#### Inspect Hits column

In [94]:
df1["Hits"].unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1600',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1.1K', '433', '242', '206', '177', '1.5K', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3.2K', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1.7K', '2.3K', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1.9K', '642', '155', '323', '288', '497', '509', '79',

In [96]:
#convert string to numeric by extract method using regural expression
def convert_to_num(value):
    if isinstance(value, str):
        
        if "K" in value:
            value = value.strip('K')
            value = float(value) * 1000
            return value 
        elif 'nan' in value:
            return np.nan
        else:
            return float(value)
    else:
        return value

df1['Hits'] = df1['Hits'].apply(convert_to_num)
df1['Hits'].unique()

array([7.71e+02, 5.62e+02, 1.50e+02, 2.07e+02, 5.95e+02, 2.48e+02,
       2.46e+02, 1.20e+02, 1.60e+03, 1.30e+02, 3.21e+02, 1.89e+02,
       1.75e+02, 9.60e+01, 1.18e+02, 2.16e+02, 2.12e+02, 1.54e+02,
       2.05e+02, 2.02e+02, 3.39e+02, 4.08e+02, 1.03e+02, 3.32e+02,
       8.60e+01, 1.73e+02, 1.61e+02, 3.96e+02, 1.10e+03, 4.33e+02,
       2.42e+02, 2.06e+02, 1.77e+02, 1.50e+03, 1.98e+02, 4.59e+02,
       1.17e+02, 1.19e+02, 2.09e+02, 8.40e+01, 1.87e+02, 1.65e+02,
       2.03e+02, 6.50e+01, 3.36e+02, 1.26e+02, 3.13e+02, 1.24e+02,
       1.45e+02, 5.38e+02, 1.82e+02, 1.01e+02, 4.50e+01, 3.77e+02,
       9.90e+01, 1.94e+02, 4.03e+02, 4.14e+02, 5.93e+02, 3.74e+02,
       2.45e+02, 3.20e+03, 2.66e+02, 2.99e+02, 3.09e+02, 2.15e+02,
       2.65e+02, 2.11e+02, 1.12e+02, 3.37e+02, 7.00e+01, 1.59e+02,
       6.88e+02, 1.16e+02, 6.30e+01, 1.44e+02, 1.23e+02, 7.10e+01,
       2.24e+02, 1.13e+02, 1.68e+02, 6.10e+01, 8.90e+01, 1.37e+02,
       2.78e+02, 7.50e+01, 1.48e+02, 1.76e+02, 1.97e+02, 2.64e

In [98]:
#remove Exponential and have the values as digits
df1['Hits'] = df1['Hits'].apply(lambda x: f"{x:.0f}")
df1['Hits'].unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1600',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1100', '433', '242', '206', '177', '1500', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3200', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1700', '2300', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1900', '642', '155', '323', '288', '497', '509', '79',

In [100]:
#convert the hits column to a float because of Nan
df1['Hits'] = df1['Hits'].astype('float')

<h3>Process to covert hits to values</h3>
<p>1. create a function to convert the values to a number <br>
by removing character 'K' and multiply it by a 1000 <br>
2. replaced nan with numpy NANs <br>
3. removed the exponential from the values</p>

<h3>Save the cleaned data as a CSV File to the current working directorate</h3>

In [103]:
#save clean data as a csv
df1.to_csv("Fifa_cleaned_data.csv")

<h2>Conclusion</h2>
<h3>In this project we will walk through the data cleaning process. <br>
The objective is to prepare the data for analysis by removing inconsistencies <br>
and converting it into a more usable format. The following key tasks were undertaken:</h3>
<h3>1. Make a copy of your DataFrame</h3>
<p>Make a copy of your data so as to retain an original copy</p>
<h3>2. Remove Unnecessary Columns</h3>
<p>Remove columns that are not significant to your analysis</p>
<h3>3. Rename columns</h3>
<p>Rename columns so as to have correct column names that can be understood</p>
<h3>4. Check for duplicate values</h3>
<p>Check for duplicated rows in your dataset</p>
<h3>5. Convert Data types of Height and Weight columns</h3>
<p>The data types of the "Height" and "Weight" columns were converted, and heights<br>
were transformed from feet (ft) to centimeters (cm), while weights were converted<br>
from pounds (lbs) to kilograms (kg).</p>
<h3>6. Clean the Contract Column</h3>
<P>Created a function to clean contract column</P>
<h3>7. Converting Date Strings into Datetime Format</h3>
<p>We converted date strings into the datetime data type, making it easier to <br>
work with time-related information.</p>
<h3>8. Converting datatypes of Value, Wage, Release Clause columns</h3>
<p>Transformed "Value", "Wage", "Release Clause" columns to int data types</p>
<h3>9. Remove '★' from "Weak Foot", "Skill Move", "International Reputation"</h3>
<p>Removed "★" character from the columns "Weak Foot", "Skill Move", "International Reputation"</p>
<h3>10. Save the Data Frame</h3>
<p>The cleaned data was saved as a CSV File</p>

<P>In this project, we embarked on a comprehensive data cleaning and transformation<br>
journey with the FIFA dataset. Our primary objective was to prepare the data for <br>
analysis by addressing inconsistencies, refining data types, and enhancing its usability.<br>
In conclusion, data cleaning and transformation are foundational steps in any data analysis project.<br>
By addressing inconsistencies, refining data types, and enhancing data quality, we have set the stage<br>
for more meaningful and insightful analyses. The clean and structured dataset is now well-equipped for<br>
advanced analytics, visualizations, and modeling.</P>